### PR TITLE
Support WeChat as notifier

### PR DESCRIPTION
**Pull request by loganhz**
Now we don't support WeChat as a notifier. So user can not send the alert to wechat. We need to support it

https://github.com/rancher/rancher/issues/17121

Related PR
https://github.com/rancher/types/pull/658

### CODE REVIEW

File: pkg/controllers/user/alert/config/notifiers.go [hunk 1]
```diff
@@ -189,6 +197,34 @@ func (c *PagerdutyConfig) UnmarshalYAML(unmarshal func(interface{}) error) error
...
+	NotifierConfig `yaml:",inline" json:",inline"`
+
+	Secret  Secret `yaml:"api_secret,omitempty" json:"api_secret,omitempty"`
+	URL     string `yaml:"api_url,omitempty" json:"api_url,omitempty"`
+	Corp    string `yaml:"corp_id,omitempty" json:"corp_id,omitempty"`
+	Message string `yaml:"message,omitempty" json:"message,omitempty"`
+	Agent   string `yaml:"agent_id,omitempty" json:"agent_id,omitempty"`
+	Party   string `yaml:"to_party,omitempty" json:"to_party,omitempty"`
+	Tag     string `yaml:"to_tag,omitempty" json:"to_tag,omitempty"`
+	User    string `yaml:"to_user,omitempty" json:"to_user,omitempty"`
```
Comment (aiwantaozi) [inline pkg/controllers/user/alert/config/notifiers.go]: can we consist the golang and json field name, like both use ToUser?

File: pkg/controllers/user/alert/config/notifiers.go [hunk 2]
```diff
@@ -189,6 +197,34 @@ func (c *PagerdutyConfig) UnmarshalYAML(unmarshal func(interface{}) error) error
...
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *WechatConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultWechatConfig
+	type plain WechatConfig
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
+	}
+	if c.Secret == "" {
+		return fmt.Errorf("missing api secret in Wechat config")
+	}
```
Comment (aiwantaozi) [inline pkg/controllers/user/alert/config/notifiers.go]: Also add crop check for the wechat config ?

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 3]
```diff
@@ -18,11 +18,11 @@ import (
...
 
-	"github.com/rancher/types/apis/core/v1"
-	"github.com/rancher/types/apis/management.cattle.io/v3"
+	v1 "github.com/rancher/types/apis/core/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
 	"github.com/rancher/types/config"
 	"github.com/sirupsen/logrus"
-	"gopkg.in/yaml.v2"
+	yaml "gopkg.in/yaml.v2"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
```
Comment (aiwantaozi) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: Better not change import, if the import name change, may affect other parts.

File: pkg/notifiers/sender.go [hunk 4]
```diff
@@ -18,7 +18,7 @@ import (
 	"time"
 
 	"github.com/prometheus/common/model"
-	"github.com/rancher/types/apis/management.cattle.io/v3"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
```
Comment (aiwantaozi) [inline pkg/notifiers/sender.go]: same as before.

File: pkg/notifiers/sender.go [hunk 5]
```diff
@@ -87,6 +95,99 @@ func TestPagerduty(key, msg string) error {
 	return nil
 }
 
+func TestWechat(secret, agent, corp, receiverType, receiver, msg string) error {
```
Comment (aiwantaozi) [inline pkg/notifiers/sender.go]: https://github.com/prometheus/alertmanager/blob/master/notify/impl.go#L989 test could check here

**Comment by loganhz** [comment]
@aiwantaozi Updated Thanks!

File: pkg/notifiers/sender.go [hunk 6]
```diff
@@ -26,6 +26,15 @@ type Message struct {
 	Content string
 }
 
+type wechatToken struct {
+	AccessToken string `json:"access_token"`
+}
+
+type weChatResponse struct {
```
Comment (aiwantaozi) [inline pkg/notifiers/sender.go]: weChat -> wechat?

File: pkg/notifiers/sender.go [hunk 7]
```diff
@@ -87,6 +104,96 @@ func TestPagerduty(key, msg string) error {
...
+	q.Add("corpid", corp)
+	q.Add("corpsecret", secret)
+	req.URL.RawQuery = q.Encode()
+
+	client := &http.Client{
+		Timeout: 15 * time.Second,
+	}
+	var wechatToken wechatToken
+	resp, err := client.Do(req)
+
```
Comment (aiwantaozi) [inline pkg/notifiers/sender.go]: can we remove this blank line? The if err != nil will be the next line after you declared the err

File: pkg/notifiers/sender.go [hunk 8]
```diff
@@ -87,6 +104,96 @@ func TestPagerduty(key, msg string) error {
...
+		return err
+	}
+
+	if wechatToken.AccessToken == "" {
+		return fmt.Errorf("Invalid APISecret for CorpID. %s", corp)
+	}
+
+	if msg == "" {
+		msg = "Wechat setting validated"
+	}
```
Comment (aiwantaozi) [inline pkg/notifiers/sender.go]: Can we move them to the beginning of the the TestWechat func?

if msg == "" {
    msg = "Wechat setting validated"
}

File: pkg/notifiers/sender.go [hunk 9]
```diff
@@ -293,6 +400,19 @@ func hashKey(s string) string {
...
+
+type wechatEvent struct {
+	Party string             `json:"toparty"`
+	User  string             `json:"touser"`
+	Tag   string             `json:"totag"`
+	Agent string             `json:"agentid"`
+	Type  string             `json:"msgtype"`
+	Text  wechatEventPayload `json:"text"`
+}
+
```
Comment (aiwantaozi) [inline pkg/notifiers/sender.go]: Can you also consist the golang field and json field here?

**Comment by loganhz** [comment]
@aiwantaozi Updated

**Comment by cjellick** [comment]
LGTM. just need to update vendor

**Comment by loganhz** [comment]
@cjellick Updated. Thanks!

**Pull request merged**